packages/flutter_tools/lib/src/resident_runner.dart

```diff
@@ -243,11 +245,16 @@ class FlutterDevice {
             disableServiceAuthCodes,
           );
         } on Exception catch (e) {
-          globals.printTrace('Fail to connect to service protocol: $observatoryUri: $e');
-          if (!completer.isCompleted && !_isListeningForObservatoryUri) {
-            completer.completeError('failed to connect to $observatoryUri');
+          final bool useExistingDdsInstance = e is dds.DartDevelopmentServiceException &&
+                                             allowExistingDdsInstance &&
+                                             e.message.contains('A DDS instance is already connected');
```

Comment:
Doing this based on the message contents seems fragile. I would recommend using an error code

Reply:
I would also suggest a separate `on` clause for the `DartDevelopmentServiceException` instead of the is check.

Reply:
Sure, that makes sense. I'll add a TODO for the error code since currently `DartDevelopmentServiceException` doesn't provide one and I'd like to get this fixed.